File: hard/1665-minimum-initial-energy-to-finish-tasks/solution.java
import java.util.Arrays;

/**
 * 1665. Minimum Initial Energy to Finish Tasks
 * Time: O(n log n)
 * Space: O(1)
 */
class Solution {
    public int minimumEffort(int[][] tasks) {
        Arrays.sort(tasks, (a, b) -> (b[1] - b[0]) - (a[1] - a[0]));

        int initial = 0;
        int current = 0;

        for (int[] task : tasks) {
            int actual = task[0];
            int minimum = task[1];

            if (current < minimum) {
                int needed = minimum - current;
                initial += needed;
                current += needed;
            }

            current -= actual;
        }

        return initial;
    }
}

File: hard/1665-minimum-initial-energy-to-finish-tasks/solution.py
from typing import List


class Solution:
    def minimumEffort(self, tasks: List[List[int]]) -> int:
        """
        1665. Minimum Initial Energy to Finish Tasks
        Time: O(n log n)
        Space: O(1)
        """
        tasks.sort(key=lambda task: task[1] - task[0], reverse=True)

        initial = 0
        current = 0

        for actual, minimum in tasks:
            if current < minimum:
                needed = minimum - current
                initial += needed
                current += needed

            current -= actual

        return initial
